File: 백준/Gold/10423. 전기가 부족해/전기가 부족해.py
import sys
from heapq import heappush, heappop

input = sys.stdin.readline

n, m, k = map(int, input().split())

generator = dict.fromkeys(list(map(int, input().split())), True)
MST_parents = [i for i in range(n + 1)]
path_parents = [i for i in range(n + 1)]
graph = []
path = []

for u, v, w in [list(map(int, input().split())) for _ in range(m)]:
    heappush(graph, (w, u, v))
    

def find_parent(x, parents):
    global generator

    if parents[x] == x:
        return x

    parents[x] = find_parent(parents[x], parents)

    return parents[x]

    
def union(px, py, parents):
    global generator

    parents[py] = px


def find_parent2(x, parents):
    global generator

    if parents[x] == x or generator.get(x) is not None:
        return x

    parents[x] = find_parent(parents[x], parents)

    return parents[x]

    
def union2(px, py, parents):
    global generator

    if generator.get(px) is not None and generator.get(py) is not None:
        return

    if px == py:
        return

    if generator.get(py) is not None:
        py, px = px, py
    
    parents[py] = px

    
answer = 0
while graph:
    cost, sour, des = heappop(graph)
    
    sour_parent, des_parent = find_parent(sour, MST_parents), find_parent(des, MST_parents)
    sour_path_parent, des_path_parent = find_parent2(sour, path_parents), find_parent2(des, path_parents)

    if sour_parent == des_parent:
        continue
    
    heappush(path, (-cost, sour, des))
    union2(sour_path_parent, des_path_parent, path_parents)
    MST_parents[sour_parent] = des_parent
    answer += cost


while path:
    cost, sour, des = heappop(path)
    
    py, px = find_parent2(sour, path_parents), find_parent2(des, path_parents)
    
    if py == px:
        continue
    answer += cost
        
    
print(answer)

File: 백준/Gold/10423. 전기가 부족해/전기가 부족해.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

class Main{

    private static int n, m;
    private static int[] parents;
    private static Map<Integer, Boolean> generators;
    private static PriorityQueue<int[]> pq;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer stk = new StringTokenizer(br.readLine());

        n = Integer.parseInt(stk.nextToken());
        m = Integer.parseInt(stk.nextToken());
        stk.nextToken();

        parents = new int[n + 1];
        for(int i=1 ; i<=n ; i++){
            parents[i] = i;
        }

        stk = new StringTokenizer(br.readLine());
        generators = new HashMap<>();
        while(stk.hasMoreTokens()){
            generators.put(Integer.parseInt(stk.nextToken()), true);
        }
        
        pq = new PriorityQueue<>((o1, o2) -> o1[2] - o2[2]);
        while(m-- > 0){
            stk = new StringTokenizer(br.readLine());
            int sour, des, cost;

            sour = Integer.parseInt(stk.nextToken());
            des = Integer.parseInt(stk.nextToken());
            cost = Integer.parseInt(stk.nextToken());

            pq.add(new int[]{sour, des, cost});
        }

        System.out.println(mst());
    }

    private static int mst(){
        int ret = 0;

        while(!pq.isEmpty()){
            int[] cur = pq.poll();
            int sour = cur[0], des = cur[1], cost = cur[2];

            int sourParent = findParent(sour), desParent = findParent(des);
            if((generators.containsKey(sourParent) && generators.containsKey(desParent))
             || sourParent == desParent)
                continue;
            
            union(sour, des);
            ret += cost;
        }

        return ret;
    }

    private static int findParent(int x){
        if(x == parents[x] || generators.containsKey(x)){
            return x;
        }

        return parents[x] = findParent(parents[x]);
    }

    private static void union(int x, int y){
        int px = findParent(x), py = findParent(y);

        if(px == py){
            return;
        }

        if(generators.containsKey(py)){
            int temp = py;
            py = px;
            px = temp;
        }

        parents[py] = px;
    }
}